# Chapter 2: Python Modules and Packages

A module is a collection of related Python objects, such as functions, classes,
and constants. Packages organize modules into larger toolboxes. This chapter
focuses on importing, inspecting, installing, and using packages responsibly.

## Learning map

1. Why modules are needed
2. Import styles and conventional aliases
3. Package metadata and object inspection
4. Help and documentation
5. Package installation commands
6. NumPy arrays
7. A modular Black-Scholes example

**Code availability:** Yes. Installation commands are shown as Markdown because
they should be run in a terminal, while Python demonstrations are independently
runnable.

## How to Study This Chapter

- Practice all import styles and identify the namespace created by each one.
- Inspect one package with version metadata, `dir()`, and its docstring.
- Separate terminal installation commands from Python code.
- Re-run every example in a fresh kernel to confirm that dependencies are local.

## Exam Checklist

- Can I write the correct import before calling a function?
- Can I explain why conventional aliases improve readability?
- Can I distinguish package installation from package import?
- Can I use a NumPy array for vectorized financial calculations?


## 2.1 Why Modules Are Needed


Python does not place every possible function in the default namespace.
Importing a module makes a focused set of tools available and keeps their origin
visible. For example, `sqrt` belongs to the standard-library `math` module.



### Structured Review

- **Namespace:** Determine whether the function is called through a module name, an alias, or a direct import.
- **Dependency:** Import only the package required by the current code block.
- **Inspection:** Use version metadata, `dir()`, and documentation to understand an unfamiliar package.
- **Numerical object:** Distinguish ordinary Python containers from vectorized NumPy arrays.

### Exam Focus

- Complete the correct import statement before using a function.
- Recognize standard aliases such as `np`, `pd`, `yf`, and `plt`.
- Explain why wildcard imports make code harder to audit.

### Common Mistakes

- Installing a package into a different Python environment.
- Calling a module function without its namespace.
- Assuming a list automatically supports NumPy-style vectorized arithmetic.


In [1]:
'''
本单元演示通过模块命名空间调用函数。
使用 math.sqrt 可以清楚看到函数来源，并避免命名冲突。
'''
import math

# 模块名位于点号左侧，函数名位于点号右侧
value = math.sqrt(3)
print("Square root of 3 =", value)


Square root of 3 = 1.7320508075688772


**Result interpretation.** `math.sqrt(3)` works because the module has been imported and the function is accessed through its namespace.

### Result Checklist

- Confirm that the sign and magnitude are economically reasonable.
- Match the output to the formula, decision rule, or statistical hypothesis described above.
- If the result differs from expectation, inspect input frequency, indexing, and parameter order first.


## 2.2 Import Styles and Conventional Aliases


Common import styles include `import module`, `import module as alias`, and
`from module import object`. Explicit imports are preferred over wildcard
imports because they make dependencies and names easier to trace.



### Structured Review

- **Namespace:** Determine whether the function is called through a module name, an alias, or a direct import.
- **Dependency:** Import only the package required by the current code block.
- **Inspection:** Use version metadata, `dir()`, and documentation to understand an unfamiliar package.
- **Numerical object:** Distinguish ordinary Python containers from vectorized NumPy arrays.

### Exam Focus

- Complete the correct import statement before using a function.
- Recognize standard aliases such as `np`, `pd`, `yf`, and `plt`.
- Explain why wildcard imports make code harder to audit.

### Common Mistakes

- Installing a package into a different Python environment.
- Calling a module function without its namespace.
- Assuming a list automatically supports NumPy-style vectorized arithmetic.


In [2]:
'''
本单元比较三种常见导入方式。
示例仅使用标准库，且所有依赖都在当前单元中导入。
'''
import math
import math as m
from math import sqrt

# 完整模块名调用：来源最清楚
print("Module namespace =", math.sqrt(9))

# 别名调用：适合名称较长且频繁使用的模块
print("Alias namespace =", m.sqrt(16))

# 直接导入函数后，不需要再写模块名前缀
print("Direct function import =", sqrt(25))


Module namespace = 3.0
Alias namespace = 4.0
Direct function import = 5.0


**Result interpretation.** All three approaches calculate square roots, but module namespaces provide the clearest provenance.

### Result Checklist

- Confirm that the sign and magnitude are economically reasonable.
- Match the output to the formula, decision rule, or statistical hypothesis described above.
- If the result differs from expectation, inspect input frequency, indexing, and parameter order first.


## 2.3 Conventional Aliases

Frequently used packages have community conventions:

```python
import numpy as np
import pandas as pd
import scipy as sp
import yfinance as yf
import matplotlib.pyplot as plt
```

Aliases are not mandatory, but following conventions makes code easier for
other analysts to read. Each notebook code unit should still import only the
packages it actually uses.


## 2.4 Inspecting a Package


`dir()` lists names exposed by an object. Package metadata can report the
installed version. Inspection is useful when learning a new financial library,
but the official documentation remains the best source for detailed behavior.



### Structured Review

- **Namespace:** Determine whether the function is called through a module name, an alias, or a direct import.
- **Dependency:** Import only the package required by the current code block.
- **Inspection:** Use version metadata, `dir()`, and documentation to understand an unfamiliar package.
- **Numerical object:** Distinguish ordinary Python containers from vectorized NumPy arrays.

### Exam Focus

- Complete the correct import statement before using a function.
- Recognize standard aliases such as `np`, `pd`, `yf`, and `plt`.
- Explain why wildcard imports make code harder to audit.

### Common Mistakes

- Installing a package into a different Python environment.
- Calling a module function without its namespace.
- Assuming a list automatically supports NumPy-style vectorized arithmetic.


In [3]:
'''
本单元检查 NumPy 的版本与可用对象数量。
按需导入 numpy 和标准库 metadata，不依赖前置单元。
'''
import numpy as np
from importlib.metadata import version

# dir() 返回模块中的名称；过滤下划线开头的内部对象
public_objects = [name for name in dir(np) if not name.startswith("_")]

print("NumPy version =", version("numpy"))
print("Number of public NumPy objects =", len(public_objects))
print("Sample objects =", public_objects[:10])


NumPy version = 2.4.6
Number of public NumPy objects = 495
Sample objects = ['False_', 'ScalarType', 'True_', 'abs', 'absolute', 'acos', 'acosh', 'add', 'all', 'allclose']


**Result interpretation.** The exact version and object count depend on the local environment, so they should be checked rather than memorized.

### Result Checklist

- Confirm that the sign and magnitude are economically reasonable.
- Match the output to the formula, decision rule, or statistical hypothesis described above.
- If the result differs from expectation, inspect input frequency, indexing, and parameter order first.


## 2.5 Reading Help Without Flooding the Notebook


`help(object)` prints the full documentation, which can be very long. In a
revision notebook, the first line of the docstring is often enough to identify
the purpose before opening the complete reference.



### Structured Review

- **Namespace:** Determine whether the function is called through a module name, an alias, or a direct import.
- **Dependency:** Import only the package required by the current code block.
- **Inspection:** Use version metadata, `dir()`, and documentation to understand an unfamiliar package.
- **Numerical object:** Distinguish ordinary Python containers from vectorized NumPy arrays.

### Exam Focus

- Complete the correct import statement before using a function.
- Recognize standard aliases such as `np`, `pd`, `yf`, and `plt`.
- Explain why wildcard imports make code harder to audit.

### Common Mistakes

- Installing a package into a different Python environment.
- Calling a module function without its namespace.
- Assuming a list automatically supports NumPy-style vectorized arithmetic.


In [4]:
'''
本单元读取函数文档的第一行，避免输出过长。
np.std 用于计算标准差，是金融风险度量中的基础函数。
'''
import numpy as np

# __doc__ 保存函数文档；这里只提取第一行作为功能摘要
summary = np.std.__doc__.strip().splitlines()[0]

# np.std 默认计算总体标准差，衡量收益率的离散程度
example = np.std([0.01, -0.02, 0.03, 0.00])

print("np.std summary:", summary)
print("Example population standard deviation =", round(float(example), 6))


np.std summary: Compute the standard deviation along the specified axis.
Example population standard deviation = 0.018028


**Result interpretation.** The docstring identifies the function, and the example measures dispersion in a small return series.

### Result Checklist

- Confirm that the sign and magnitude are economically reasonable.
- Match the output to the formula, decision rule, or statistical hypothesis described above.
- If the result differs from expectation, inspect input frequency, indexing, and parameter order first.


## 2.6 Installing and Managing Packages

Run package-management commands in **Anaconda Prompt**, PowerShell, Terminal, or
a Jupyter cell prefixed with `!`. Typical commands are:

```text
python -m pip install numpy
python -m pip install yfinance
python -m pip install numpy-financial
python -m pip install --upgrade pip
python -m pip list
python -m pip show numpy
```

Version constraints can improve reproducibility:

```text
python -m pip install "SomeProject==1.4"
python -m pip install "SomeProject>=1,<2"
```

Use the active environment's Python executable. Installing into a different
environment is a common reason that an import still fails after installation.


## 2.7 NumPy Arrays for Financial Data


A NumPy array stores homogeneous numerical data and supports vectorized
operations. It is more suitable than a plain list for repeated numerical
calculations such as return transformations and scenario analysis.



### Structured Review

- **Namespace:** Determine whether the function is called through a module name, an alias, or a direct import.
- **Dependency:** Import only the package required by the current code block.
- **Inspection:** Use version metadata, `dir()`, and documentation to understand an unfamiliar package.
- **Numerical object:** Distinguish ordinary Python containers from vectorized NumPy arrays.

### Exam Focus

- Complete the correct import statement before using a function.
- Recognize standard aliases such as `np`, `pd`, `yf`, and `plt`.
- Explain why wildcard imports make code harder to audit.

### Common Mistakes

- Installing a package into a different Python environment.
- Calling a module function without its namespace.
- Assuming a list automatically supports NumPy-style vectorized arithmetic.


In [5]:
'''
本单元使用 NumPy 数组保存资产收益率并进行向量化计算。
所有收益率同时乘以投资金额，得到各情景下的盈亏。
'''
import numpy as np

returns = np.array([0.02, -0.01, 0.015, 0.005])
investment = 10_000

# 数组与标量相乘会逐元素运算，无需手写 for 循环
profit_and_loss = investment * returns

print("Returns =", returns)
print("Scenario P&L =", profit_and_loss)
print("Average return =", round(float(returns.mean()), 6))


Returns = [ 0.02  -0.01   0.015  0.005]
Scenario P&L = [ 200. -100.  150.   50.]
Average return = 0.0075


**Result interpretation.** Vectorization converts all return scenarios into dollar profit and loss without writing an explicit loop.

### Result Checklist

- Confirm that the sign and magnitude are economically reasonable.
- Match the output to the formula, decision rule, or statistical hypothesis described above.
- If the result differs from expectation, inspect input frequency, indexing, and parameter order first.


## 2.8 Modular Black-Scholes Call Function


The lecture uses the Black-Scholes call formula to demonstrate that a function
can import only the objects it needs. The example also previews how NumPy and
SciPy support later finance topics.

$$C=S\Phi(d_1)-Xe^{-rT}\Phi(d_2)$$



### Structured Review

- **Object:** Identify the value and data type stored by each variable.
- **Access:** Remember that Python sequence indexing begins at zero.
- **Mutability:** Decide whether the object may be changed after creation.
- **Function contract:** Separate parameters, calculation steps, and the returned value.

### Exam Focus

- Preserve indentation inside functions and conditional blocks.
- Distinguish positional arguments from keyword arguments.
- Trace the value returned by a function for a small input.

### Common Mistakes

- Using the wrong capitalization in a variable name.
- Attempting to modify a tuple.
- Printing a result inside a function when the question requires `return`.


In [6]:
'''
本单元实现欧式看涨期权的 Black-Scholes 定价函数。
函数内部按需导入 NumPy 数学函数与 SciPy 正态分布工具，
因此本单元可独立运行，并清楚展示模块化依赖。
'''
def black_scholes_call(S, X, T, r, sigma):
    # 函数内部按需导入：log、exp、sqrt 负责公式计算
    from numpy import exp, log, sqrt

    # norm.cdf() 计算标准正态分布的累积概率 Phi(d)
    from scipy.stats import norm

    # d1 同时包含价内程度、无风险利率、波动率和到期时间
    d1 = (log(S / X) + (r + sigma**2 / 2) * T) / (sigma * sqrt(T))

    # d2 等于 d1 减去一个波动率时间尺度
    d2 = d1 - sigma * sqrt(T)

    # 看涨期权价值 = 股票现值部分 - 行权价格现值部分
    return S * norm.cdf(d1) - X * exp(-r * T) * norm.cdf(d2)

call_value = black_scholes_call(S=100, X=100, T=1, r=0.05, sigma=0.20)
print("European call value =", round(float(call_value), 4))


European call value = 10.4506


**Result interpretation.** For the stated assumptions, the model produces a call value of approximately $10.45.

### Result Checklist

- Confirm that the sign and magnitude are economically reasonable.
- Match the output to the formula, decision rule, or statistical hypothesis described above.
- If the result differs from expectation, inspect input frequency, indexing, and parameter order first.


## Chapter Review

- Modules organize related programs and reduce duplication.
- Prefer explicit imports and standard aliases.
- Use `dir()`, metadata, docstrings, and official documentation to inspect tools.
- Install packages into the same environment that runs the notebook.
- NumPy enables vectorized numerical analysis.
- Functions should declare their own required dependencies when independence is important.
